### ISSUE TITLE
Incorrect diagnosis for into trait 

### ISSUE THREAD

**Issue by rust1248** [intellij-rust/intellij-rust#9108]
## Environment

* **IntelliJ Rust plugin version:** 0.4.174.4743-221
* **Rust toolchain version:** 1.59.0
* **IDE name and version:** CLion 2022.1.2
* **Operating system:** Ubuntu 18.04

## Problem description
Into incorrectly detect recursive function and errors with mismatched types on this code:

```rust
struct Test {
    s: String,
}

impl Test {
    fn new(s: String) -> Test {
        Test { s }
    }
}

impl<'a> Into<Test> for &'a str {
    fn into(self) -> Test {
        Test::new(self.into())  //  <- E0308 expected String found Test
    }
}

// included main to easy copy-paste
fn main() {}
```

but it compiles fine. The into call is also displayed as a recursive call when it shouldn't.

## Steps to reproduce

Copy the code and see
